python/cudf/cudf/core/indexed_frame.py

```diff
@@ -414,3 +415,55 @@ def sort_index(
         if ignore_index is True:
             out = out.reset_index(drop=True)
         return self._mimic_inplace(out, inplace=inplace)
+
+    def _gather(self, gather_map, keep_index=True, nullify=False):
+        gather_map = cudf.core.column.as_column(gather_map)
+        if not is_integer_dtype(gather_map.dtype):
+            gather_map = gather_map.astype("int32")
```

Comment:
If this kind of pattern pops up a lot, enforcing array requirements can be refactored into a helper method that uses a declarative style. For example:

```python
values = util._convert_array(values, dtype=np.int32)

# or 

values = util._as_integer_dtype(values)
```

Here is an example of such a method. https://github.com/glotzerlab/freud/blob/78cde335b718c78a0ec78d2ea867d2f176a5ed1e/freud/util.pyx#L199-L200

This function passes through the input untouched (zero-copy) if the array meets the requirements (which might also include things like having contiguous flags or a particular shape) and converts (introducing a copy) or raises an error if invalid.

This conversation can be resolved with no action, or we can file an issue if you think this is a common pattern that we could improve in cuDF.

Reply:
On the one hand, a function like this can really come in handy. On the other, I'm really wary that it may be abused like `as_column`. Overall I think we should encourage developers to perform conversions that works for their specific use case and perform optimization if possible. Only when developer is unable to determine input dtypes (which should only happen at very surface levels) that it should use a generic conversion method.

A premise to this is to generalize calling conventions on internal python helpers. In this scenario, I'd rather have users of `_gather` to convert `gather_map` into integral types before passing them in `_gather`, since they usually have more information what `gather_map` looks like in its use case. In general we should limit the number of internal functions that accepts `Any` dtype inputs and add type annotations and assertions to internals.

cc @vyasr @shwina

Reply:
Yeah I'm happy to make that issue.

Reply:
Documented in #9593